File: packages/a2a-sdk/a2a_sdk/card.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

import httpx

AGENT_CARD_PATHS = (
    "/.well-known/agent-card.json",
    "/.well-known/agent.json",
)


@dataclass
class AgentSkill:
    id: str
    name: str
    description: str = ""
    tags: list[str] = field(default_factory=list)
    examples: list[str] = field(default_factory=list)
    input_modes: list[str] = field(default_factory=list)
    output_modes: list[str] = field(default_factory=list)

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> AgentSkill:
        return cls(
            id=raw.get("id") or raw.get("skill_id") or "",
            name=raw.get("name") or "",
            description=raw.get("description") or "",
            tags=list(raw.get("tags") or []),
            examples=list(raw.get("examples") or []),
            input_modes=list(raw.get("inputModes") or raw.get("input_modes") or []),
            output_modes=list(raw.get("outputModes") or raw.get("output_modes") or []),
        )


@dataclass
class AgentCard:
    name: str
    description: str
    url: str
    version: str
    protocol_version: str = "0.3.0"
    capabilities: dict[str, Any] = field(default_factory=dict)
    default_input_modes: list[str] = field(default_factory=lambda: ["text"])
    default_output_modes: list[str] = field(default_factory=lambda: ["text"])
    skills: list[AgentSkill] = field(default_factory=list)
    raw: dict[str, Any] = field(default_factory=dict)

    def skill_ids(self) -> list[str]:
        return [s.id for s in self.skills if s.id]

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> AgentCard:
        skills = [AgentSkill.from_dict(s) for s in raw.get("skills", [])]
        return cls(
            name=raw.get("name") or "",
            description=raw.get("description") or "",
            url=raw.get("url") or "",
            version=raw.get("version") or "0.0.0",
            protocol_version=str(raw.get("protocolVersion") or raw.get("protocol_version") or "0.3.0"),
            capabilities=dict(raw.get("capabilities") or {}),
            default_input_modes=list(
                raw.get("defaultInputModes") or raw.get("default_input_modes") or ["text"]
            ),
            default_output_modes=list(
                raw.get("defaultOutputModes") or raw.get("default_output_modes") or ["text"]
            ),
            skills=skills,
            raw=raw,
        )


def _normalize_base(url: str) -> str:
    return url.rstrip("/")


def fetch_agent_card(base_url: str, timeout: float = 10.0) -> AgentCard:
    """Fetch Agent Card from well-known paths (canonical then legacy)."""
    base = _normalize_base(base_url)
    errors: list[str] = []

    with httpx.Client(timeout=timeout, follow_redirects=True) as client:
        for path in AGENT_CARD_PATHS:
            url = f"{base}{path}"
            try:
                resp = client.get(url)
                if resp.status_code == 404:
                    errors.append(f"{path}: 404")
                    continue
                resp.raise_for_status()
                data = resp.json()
                card = AgentCard.from_dict(data)
                if not card.url:
                    card.url = f"{base}/"
                return card
            except Exception as exc:  # noqa: BLE001 — collect and raise aggregate
                errors.append(f"{path}: {exc}")

    raise LookupError(
        f"Failed to fetch Agent Card from {base}. Tried {list(AGENT_CARD_PATHS)}. "
        f"Errors: {'; '.join(errors)}"
    )

File: apps/gateway/internal/registry/card.go
package registry

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
	"unicode"
)

type AgentCard struct {
	Name               string         `json:"name"`
	Description        string         `json:"description"`
	URL                string         `json:"url"`
	Version            string         `json:"version"`
	ProtocolVersion    string         `json:"protocolVersion"`
	Capabilities       map[string]any `json:"capabilities"`
	DefaultInputModes  []string       `json:"defaultInputModes"`
	DefaultOutputModes []string       `json:"defaultOutputModes"`
	Skills             []Skill        `json:"skills"`
}

type Skill struct {
	ID          string   `json:"id"`
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Tags        []string `json:"tags"`
	Examples    []string `json:"examples"`
	InputModes  []string `json:"inputModes"`
	OutputModes []string `json:"outputModes"`
}

type CardFetcher struct {
	Client *http.Client
}

func NewCardFetcher(timeout time.Duration) *CardFetcher {
	return &CardFetcher{
		Client: &http.Client{Timeout: timeout},
	}
}

func (f *CardFetcher) Fetch(ctx context.Context, endpoint string) (AgentCard, []byte, error) {
	base, err := normalizeBase(endpoint)
	if err != nil {
		return AgentCard{}, nil, err
	}

	paths := []string{
		"/.well-known/agent-card.json",
		"/.well-known/agent.json",
	}

	var lastErr error
	for _, p := range paths {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, base+p, nil)
		if err != nil {
			lastErr = err
			continue
		}
		resp, err := f.Client.Do(req)
		if err != nil {
			lastErr = err
			continue
		}
		body, readErr := io.ReadAll(resp.Body)
		_ = resp.Body.Close()
		if readErr != nil {
			lastErr = readErr
			continue
		}
		if resp.StatusCode == http.StatusNotFound {
			lastErr = fmt.Errorf("%s: 404", p)
			continue
		}
		if resp.StatusCode >= 300 {
			lastErr = fmt.Errorf("%s: status %d", p, resp.StatusCode)
			continue
		}
		var card AgentCard
		if err := json.Unmarshal(body, &card); err != nil {
			lastErr = fmt.Errorf("%s: invalid json: %w", p, err)
			continue
		}
		if card.URL == "" {
			card.URL = base + "/"
		}
		if card.Version == "" {
			card.Version = "0.0.0"
		}
		return card, body, nil
	}
	if lastErr == nil {
		lastErr = fmt.Errorf("agent card not found")
	}
	return AgentCard{}, nil, lastErr
}

func (f *CardFetcher) Health(ctx context.Context, endpoint string) error {
	base, err := normalizeBase(endpoint)
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, base+"/health", nil)
	if err != nil {
		return err
	}
	resp, err := f.Client.Do(req)
	if err != nil {
		// fallback: card endpoint proves reachability
		_, _, cardErr := f.Fetch(ctx, endpoint)
		return cardErr
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		_, _, cardErr := f.Fetch(ctx, endpoint)
		return cardErr
	}
	return nil
}

func normalizeBase(endpoint string) (string, error) {
	endpoint = strings.TrimSpace(endpoint)
	if endpoint == "" {
		return "", fmt.Errorf("endpoint is required")
	}
	if !strings.Contains(endpoint, "://") {
		endpoint = "http://" + endpoint
	}
	u, err := url.Parse(endpoint)
	if err != nil {
		return "", fmt.Errorf("invalid endpoint: %w", err)
	}
	if u.Scheme == "" || u.Host == "" {
		return "", fmt.Errorf("invalid endpoint url")
	}
	u.Path = strings.TrimRight(u.Path, "/")
	u.RawQuery = ""
	u.Fragment = ""
	return strings.TrimRight(u.String(), "/"), nil
}

func DeriveAgentKey(card AgentCard, endpoint string) string {
	if card.Name != "" {
		return slugify(card.Name)
	}
	base, err := normalizeBase(endpoint)
	if err != nil {
		return "agent"
	}
	u, _ := url.Parse(base)
	if u != nil && u.Hostname() != "" {
		return slugify(u.Hostname())
	}
	return "agent"
}

func slugify(s string) string {
	s = strings.ToLower(strings.TrimSpace(s))
	var b strings.Builder
	prevDash := false
	for _, r := range s {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			b.WriteRune(r)
			prevDash = false
			continue
		}
		if !prevDash {
			b.WriteByte('-')
			prevDash = true
		}
	}
	out := strings.Trim(b.String(), "-")
	if out == "" {
		return "agent"
	}
	return out
}
